### PR TITLE
adding updating view

### PR DESCRIPTION
**Pull request by marinaRomanchuk**

### CODE REVIEW

**Comment by revimi** [comment]
Please add `.idea` to `.gitignore`

File: users/views.py [hunk 1]
```diff
@@ -1,9 +1,27 @@
...
+class GetView(generics.RetrieveAPIView):
+    queryset = User.objects.all()
+    permission_classes = (IsAuthenticated,)
+    serializer_class = UserPrivateSerializer
+    lookup_field = "username"
+
+
+class UpdateView(generics.UpdateAPIView):
+    queryset = User.objects.all()
+    permission_classes = (IsAuthenticated,)
```
Comment (revimi) [inline users/views.py:22]: We need object permissions so that user can update only their own profile

File: users/views.py [hunk 2]
```diff
@@ -1,9 +1,27 @@
...
+    serializer_class = UpdateSerializer
+    lookup_field = "username"
```
Comment (revimi) [inline users/views.py:24]: What is it for?

File: users/views.py [hunk 3]
```diff
@@ -1,9 +1,27 @@
...
 
 
 class SignupView(generics.CreateAPIView):
     permission_classes = (AllowAny,)
     serializer_class = SignupSerializer
+
+
+class GetView(generics.RetrieveAPIView):
+    queryset = User.objects.all()
+    permission_classes = (IsAuthenticated,)
```
Comment (revimi) [inline users/views.py]: We need object permissions so that user can update only their own profile

File: users/views.py (hunk 2 above)
Comment (revimi) [inline users/views.py]: What is it for?

File: users/urls.py [hunk 4]
```diff
@@ -1,7 +1,10 @@
+from django.conf.urls import url
 from django.urls import path
 
-from users.views import SignupView
+from users.views import GetView, SignupView, UpdateView
 
 urlpatterns = [
     path("signup/", SignupView.as_view()),
+    url(r"^(?P<username>\w+)/edit/$", UpdateView.as_view(), name="update"),
```
Comment (revimi) [inline users/urls.py]: We usually have one endpoint for getting, updating and removing stuff. Just use different HTTP methods for different actions

File: users/serializers.py [hunk 5]
```diff
@@ -54,3 +54,15 @@ def create(self, validated_data):
         user.save()
 
         return user
+
+
+class UpdateSerializer(serializers.ModelSerializer):
```
Comment (revimi) [inline users/serializers.py]: Why do we need a separate serializer?

File: users/views.py [hunk 6]
```diff
@@ -1,9 +1,27 @@
...
+from users.models import User
+from users.serializers import SignupSerializer, UpdateSerializer, UserPrivateSerializer
 
 
 class SignupView(generics.CreateAPIView):
     permission_classes = (AllowAny,)
     serializer_class = SignupSerializer
+
+
+class GetView(generics.RetrieveAPIView):
```
Comment (revimi) [inline users/views.py]: We have [RetrieveUpdateAPIView](https://www.django-rest-framework.org/api-guide/generic-views/#retrieveupdateapiview)

File: users/urls.py (hunk 4 above)
Comment (revimi) [reply users/urls.py]: Please use user id instead of username in the url

File: users/views.py [hunk 7]
```diff
@@ -1,9 +1,27 @@
...
+    lookup_field = "username"
+
+
+class UpdateView(generics.UpdateAPIView):
+    queryset = User.objects.all()
+    permission_classes = (IsAuthenticated,)
+    serializer_class = UpdateSerializer
+    lookup_field = "username"
+
+    def get_object(self):
```
Comment (revimi) [inline users/views.py]: You have a username parameter in the url, but here you pass the current user as an object. I understand your idea, but it is not the way we usually do it

**Review by revimi — CHANGES_REQUESTED** [review]

File: users/views.py (hunk 3 above)
Comment (marinaRomanchuk) [reply users/views.py]: Done

File: users/views.py (hunk 2 above)
Comment (marinaRomanchuk) [reply users/views.py]: It doesn't work without   lookup_field="id"

File: users/urls.py (hunk 4 above)
Comment (marinaRomanchuk) [reply users/urls.py]: Done

File: users/serializers.py (hunk 5 above)
Comment (marinaRomanchuk) [reply users/serializers.py]: Removed

File: users/views.py (hunk 6 above)
Comment (marinaRomanchuk) [reply users/views.py]: Changed to RetrieveUpdateAPIView

File: users/views.py (hunk 7 above)
Comment (marinaRomanchuk) [reply users/views.py]: Function was removed

File: users/views.py [hunk 8]
```diff
@@ -1,9 +1,18 @@
...
 class SignupView(generics.CreateAPIView):
     permission_classes = (AllowAny,)
     serializer_class = SignupSerializer
+
+
+class UpdateProfileView(generics.RetrieveUpdateAPIView):
+    queryset = User.objects.all()
+    permission_classes = (IsOwner,)
+    serializer_class = UserPrivateSerializer
+    lookup_field = "id"
```
Comment (revimi) [inline users/views.py]: Please remove this line as PK (primary key which is ID in our case) is the lookup field by default. Just update `id` to `pk` in the url

File: users/serializers.py [hunk 9]
```diff
@@ -17,6 +17,8 @@ class Meta:
 
 
 class UserPrivateSerializer(UserPublicSerializer):
+    username = serializers.CharField(write_only=True)
```
Comment (revimi) [inline users/serializers.py]: We need it to be read only not write only I suppose

File: users/urls.py [hunk 10]
```diff
@@ -1,7 +1,9 @@
+from django.conf.urls import url
 from django.urls import path
 
-from users.views import SignupView
+from users.views import SignupView, UpdateProfileView
 
 urlpatterns = [
     path("signup/", SignupView.as_view()),
+    url(r"^(?P<id>\d+)/$", UpdateProfileView.as_view()),
```
Comment (revimi) [inline users/urls.py]:
```suggestion
    url(r"^(?P<pk>\d+)/$", UpdateProfileView.as_view()),
```

File: users/views.py (hunk 8 above)
Comment (revimi) [reply users/views.py]: Also please add something like this to manage data returned by the endpoint: 
```
    def get_serializer_class(self):
        # todo cache object to not query the database twice
        if self.request.user == self.get_object():
            return UserPrivateSerializer
        return UserPublicSerializer
```

File: instagram_own_version/permissions.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+from rest_framework import permissions
+
+
+class IsOwner(permissions.BasePermission):
+    def has_permission(self, request, view):
+        return request.user.is_authenticated
+
+    def has_object_permission(self, request, view, obj):
```
Comment (revimi) [inline instagram_own_version/permissions.py:8]: We do need to check if the request method is safe (ex. GET request) and allow safe requests
```
from rest_framework.permissions import SAFE_METHODS

if request.method in SAFE_METHODS:
...
```

**Review by revimi — CHANGES_REQUESTED** [review]

File: users/views.py (hunk 8 above)
Comment (marinaRomanchuk) [reply users/views.py]: Done

File: users/serializers.py (hunk 9 above)
Comment (marinaRomanchuk) [reply users/serializers.py]: Changed to read_only

File: users/urls.py (hunk 10 above)
Comment (marinaRomanchuk) [reply users/urls.py]: Done

File: instagram_own_version/permissions.py (hunk 11 above)
Comment (marinaRomanchuk) [reply instagram_own_version/permissions.py:8]: Done

File: users/urls.py [hunk 12]
```diff
@@ -1,7 +1,9 @@
+from django.conf.urls import url
 from django.urls import path
 
-from users.views import SignupView
+from users.views import SignupView, UpdateProfileView
 
 urlpatterns = [
     path("signup/", SignupView.as_view()),
+    url(r"^(?P<pk>\d+)/$", UpdateProfileView.as_view()),
```
Comment (revimi) [inline users/urls.py]: please add `users/` prefix

File: users/serializers.py [hunk 13]
```diff
@@ -25,6 +27,7 @@ class Meta(UserPublicSerializer.Meta):
             "description",
             "date_of_birth",
             "profile_photo",
+            "email",
```
Comment (revimi) [inline users/serializers.py:35]: Let's add readonly `id` field to all serializers

**Comment by revimi** [comment]
Have you tried setting a profile picture?

File: users/views.py [hunk 14]
```diff
@@ -1,9 +1,26 @@
...
+    UserPublicSerializer,
+)
 
 
 class SignupView(generics.CreateAPIView):
     permission_classes = (AllowAny,)
     serializer_class = SignupSerializer
+
+
+class UpdateProfileView(generics.RetrieveUpdateAPIView):
```
Comment (revimi) [inline users/views.py]:
```suggestion
class RetrieveUpdateUserProfileView(generics.RetrieveUpdateAPIView):
```

Comment (marinaRomanchuk) [reply users/views.py]: Done

File: users/serializers.py (hunk 13 above)
Comment (marinaRomanchuk) [reply users/serializers.py:35]: Done

File: users/urls.py (hunk 12 above)
Comment (marinaRomanchuk) [reply users/urls.py]: Done

**Review by revimi — APPROVED** [review]

**Pull request merged**